README.md
## aKFrameWork v1.012

<i>Thanks for download my library. You can use this in your project freely 
but use it on your own risk as I'll be responsible for any damage to
your pc done from my libary. I'm not saying that my libary will harm
your computer but USE IT PROPERLY.</i> ;) :)

To use this libary, see it below
<pre>
'$include:'aKFrameWork_global.bi'
[your code]
[your code]
[your code]
'$include:'aKFrameWork_method.bi'
</pre>
Yes, include aKFrameWork_global.bi on the top of your program and 
aKFrameWork_method.bi on the bottom of your program.

<u>You can check some cool demo in the 'demo' folder. Also, you can 
read aKFrameWork.pdf in 'doc' folder to know more about SUBS and 
FUNCTION of this framework and how to use it. 

The new added FrameworkManager is a Web Application which will help you to get
.bas code of the dialog.</u>

Have any question about this libary then contact me at http://www.qb64.net/forum
Have any suggestion/bugs for this libary then inform me at
http://www.qb64.net/forum

<hr>


CHANGES.txt
v1.012
- Added Drag feature to dialog boxes (suggested by Fellippe).
- Reduce Mouse Buffer.
- Otimized code for smooth animation and low CPU usage.
- Now use _FREEIMAGE to free temporary handle of the images.
- Fixed aKTextBox() bug (reported by jc3bull).
- Added FrameworkManager.html which is used generate code of the dialog.

 


demo_form.bas
IF _DIREXISTS("akframework") THEN
    CHDIR "akframework\demo"
ELSEIF _DIREXISTS("demo") THEN
    CHDIR "demo"
END IF

'$include:'akframework\akframework_global.bi'

SCREEN _NEWIMAGE(800, 700, 32)
'setting background
_PUTIMAGE , _LOADIMAGE("bg.jpg")
_TITLE "Registration Form Demo"
'creating registration form dialog
regForm = aKNewdialog("Please fill the form", 350, 250)
'adding textboxes which will take first-name, last-name,password no.
'and email address
firstName = aKAddTextBox(regForm, "", "First Name", 10, 10, 160, 0)
lastName = aKAddTextBox(regForm, "", "Last Name", 180, 10, 160, 0)
email = aKAddTextBox(regForm, "", "Email Address", 10, 40, 330, 0)
password = aKAddTextBox(regForm, "", "Your Password", 10, 70, 330, aKPassword)
'adding  radio button for the person
label1 = aKAddlabel(regForm, "Choose Gender", 10, 110)
male = aKAddRadioButton(regForm, "Male", 140, 110, 1)
female = aKAddRadioButton(regForm, "Female", 200, 110, 1)
aKSetRadioValue regForm, male, 1
'adding a combo box to choose a city
city = aKAddComboBox(regForm, "Select City", "London,New York,Delhi,Sydney,Germany", 220, 135)
'adding numeric up-down for age.
label2 = aKAddlabel(regForm, "Age", 10, 138)
age = aKAddNumericUpDown(regForm, 22, 70, 132, 24)
'and finally adding submit,reset and cancel button
subBtn = aKAddButton(regForm, "Submit", 5, 180)
resetBtn = aKAddButton(regForm, "Reset", 75, 180)
cancelBtn = aKAddButton(regForm, "Cancel", 140, 180)
DO
    'always use this two subs at the top of the loop with the dialog handle
    aKCheck regForm
    aKUpdate regForm
    'checking if user click cancel button
    IF aKClick(regForm, aKButton, cancelBtn) THEN
        aKHideDialog regForm
        END
    END IF
    'checking if user click reset button
    IF aKClick(regForm, aKButton, resetBtn) THEN
        'we'll reset values of all objects
        aKSetValue aKTextBox, firstName, ""
        aKSetValue aKTextBox, lastName, ""
        aKSetValue aKTextBox, email, ""
        aKSetValue aKTextBox, password, ""
        aKSetRadioValue regForm, male, 1
        aKSetValue aKComboBox, city, "Select City"
        aKSetValue aKNumericUpDown, age, "22"
        'update everything.
        aKDrawObject regForm, aKTextBox, firstName
        aKDrawObject regForm, aKTextBox, lastName
        aKDrawObject regForm, aKTextBox, email
        aKDrawObject regForm, aKTextBox, password
        aKDrawObject regForm, aKComboBox, city
        aKDrawObject regForm, aKNumericUpDown, age
        aKDrawObject regForm, aKRadioButton, male
        aKDrawObject regForm, aKRadioButton, female
    END IF
    'checking if user hit submit btn
    IF aKClick(regForm, aKButton, subBtn) THEN enter = -1: EXIT DO
    _DISPLAY
    'checking if the user close the dialog
    IF aKDialogClose(regForm) THEN SYSTEM
LOOP

IF NOT enter THEN END
SCREEN 0
PRINT "Fist name : "; aKGetValue(aKTextBox, firstName)
PRINT "Last name : "; aKGetValue(aKTextBox, lastName)
PRINT "Email     : "; aKGetValue(aKTextBox, email)
PRINT "Password  : "; aKGetValue(aKTextBox, password)
PRINT "City      : "; aKGetValue(aKComboBox, city)
PRINT "Gender    : "; aKGetRadioValue(regForm, 1)
PRINT "Age       : "; aKGetValue(aKNumericUpDown, age)

 
'$include:'akframework\akframework_method.bi'


demo/demo_tooltip.bas
IF _DIREXISTS("akframework") THEN
    CHDIR "akframework\demo"
ELSEIF _DIREXISTS("demo") THEN
    CHDIR "demo"
END IF
ON ERROR GOTO getError
'$include:'akframework\akframework_global.bi'


SCREEN _NEWIMAGE(700, 600, 32)
CLS

'creating our dialog handle.
dialog = aKNewdialog("Tooltip Demo", 400, 250)
'for our demo, we'll create 5 objects
'- 2 picture boxes
'- 1 button
'- 1 label
'- 1 link label
'You can use tooltip only on some objects
'listed above, i.e. picture boxes, button,
'label and link label.

'adding picture boxes
pic1 = aKAddPicture(dialog, 10, 10, 200, 200, _LOADIMAGE("bg.jpg"))
pic2 = aKAddPicture(dialog, 220, 10, 170, 30, _LOADIMAGE("colorpicker.png"))
'a label
label = aKAddlabel(dialog, "This is a label", 220, 60)
'link and button
link = aKAddLinklabel(dialog, "Link Label1", 220, 100)
button = aKAddButton(dialog, "Hover on Me!!", 220, 140)
'now, we'll add tooltips to above objects by
'using a function aKAddTooltip. To use this, here its
'syntax -
'myTooltip% = aKAddTooltip(handle%,typ%,object%,text$)
'in which
'handle% : the handle of your dialog.
'typ% : the type of the object given. Example -
'aKLabel, aKLinkLabel, aKButton and aKPicture
'object% : the handle of the object in which you
'want to add tooltip
'text$ : the text which will be shown when the object
'will be hovered by the mouse.
tooltip1 = aKAddTooltip(dialog, aKPicture, pic1, "This tooltip is on image!!.")
tooltip2 = aKAddTooltip(dialog, aKPicture, pic2, "This tooltip on another image!!.")
'NOTE : To break lines in tooltip use '\n' with text$
'You can see a example below -
tooltip3 = aKAddTooltip(dialog, aKLabel, label, "This is line 1\nThis is line 2")
tooltip4 = aKAddTooltip(dialog, aKLinkLabel, link, "Tooltip on links!!")
tooltip5 = aKAddTooltip(dialog, aKButton, button, "Tooltips!! even on buttons")
'if my framework is like by you then please
'comment what you feels at http://www.qb64.net/forum

DO
    aKCheck dialog
    aKUpdate dialog
    _DISPLAY
    IF aKDialogClose(dialog) THEN SYSTEM
LOOP


getError:
IF ERR = 258 THEN
    SCREEN 0
    PRINT "Images file not found"
END IF

'$include:'akframework\akframework_method.bi'


demo/demo_animation.bas
IF _DIREXISTS("akframework") THEN
    CHDIR "akframework\demo"
ELSEIF _DIREXISTS("demo") THEN
    CHDIR "demo"
END IF

'$include:'akframework\akframework_global.bi'

SCREEN _NEWIMAGE(800, 700, 32)

_PUTIMAGE (0, 0), _LOADIMAGE("bg.jpg") 'sets the background of our screen
'set programs title
_TITLE "aKFrameWork Demo"

'create new dialog handle with desire hight and width
dialog = aKNewdialog("Animation Effect Demo", 600, 100)
'adding some labels to our handle at 10,10 position under it.
labels = aKAddlabel(dialog, "Select any animation from the combobox below", 10, 10)
'adding combobox which have all animation options.
'commas are used to separate options.
combobox = aKAddComboBox(dialog, "No Effect", "No Effect,FadeBlack,FadeWhite,CrosFade,Focus,ShapeOut,Slide,Blinds", 10, 40)
'adding button for running animation
runBtn = aKAddButton(dialog, "Show Dialog", 120, 40)
'adding tooltips to our button handle and defined it's type as
'aKButton. Use '\n' for line breaks in tooltips
t = aKAddTooltip(dialog, aKButton, runBtn, "click the view dialog with\ndesire animation effect.")
'creating one more dialogs in which we will
'apply animation using aKAddTransiton
miniDialog = aKNewdialog("Mini-Dialog", 400, 200)
DO
    'use these two subs at the top of the loop with
    'your dialog handle
    aKCheck dialog
    aKUpdate dialog
    'checking if our button has been clicked by the user
    'using aKClick function. Usage :
    'result% = aKClick(handle%, type%, objectHandle%)
    'handle% : the handle of our dialog
    'type% : the type of the object given like aKLable, aKButton, aKPicture or aKLinkLabel etc
    'objectHandle% : the handle of our object
    'return -1 if the following object has been clicked, 0 if not.
    IF aKClick(dialog, aKButton, runBtn) THEN
        'first we will get the value of our combo box using akGetValue Function
        'usage :
        'result$ = aKGetValue(type%, objectHandle%)
        'type% : the type of the object given like akKTextBox, aKComboBox, aKNumericUpDown, aKCheckBox
        'objectHandle% : the handle of our object
        'return the value of the desire object
        'we are using aKAddTransition to add effect to our dialog
        'usage:
        'aKAddTransition handle%,effect$
        'handle% : the handle of our dialog
        'effect$ : can be name of the effect like fadewhite,fadeblack,focus,crosfade,slide,blinds and shapeout
        aKAddTransition miniDialog, aKGetValue(aKComboBox, combobox)
        DO
            aKCheck miniDialog
            aKUpdate miniDialog
            'checking if the dialog has been closed by the user.
            IF aKDialogClose(miniDialog) THEN EXIT DO
            _DISPLAY
        LOOP
        aKFreeDialog miniDialog
        'use it always whenever a dialog as been closed by the user
        'so that it can be reuse.
    END IF
    IF aKDialogClose(dialog) THEN SYSTEM
    _DISPLAY
LOOP
'for more info read aKFramework_uses.pdf in doc folder



'$include:'akframework\akframework_method.bi'


demo/demo_progressbar.bas
IF _DIREXISTS("akframework") THEN
    CHDIR "akframework\demo"
ELSEIF _DIREXISTS("demo") THEN
    CHDIR "demo"
END IF
'$include:'akframework\akframework_global.bi'

SCREEN _NEWIMAGE(800, 700, 32)
'setting background
_PUTIMAGE , _LOADIMAGE("bg.jpg")
'setting title
_TITLE "ProgressBar Demo"

'creating a dialog handle
progressDialog = aKNewdialog("Progress Bar Demo", 380, 150)
'creating a label
label1 = aKAddlabel(progressDialog, "Click the button below to see a", 10, 10)
label2 = aKAddlabel(progressDialog, "beautiful progressbar running.", 10, 30)
'creating a run button which we'll use to run progress bar
'No matter if you are using it directly or not.
runBtn = aKAddButton(progressDialog, "Run Progress", 10, 90)
progress = aKAddProgressBar(progressDialog, 10, 50, 340, 1, -1)
'#####################################
'How shall I create progress bar   ?
'Answer:
'Read aKFramework.pdf in doc folder or see it below
'progress% = aKAddProgressBar(handle%,xpos%,ypos%,width%,value%,active%)
'handle% : the handle of your dialog.
'xpos% : the x-position of the progress bar inside the dialog.
'ypos% : the y-position of the progress bar inside the dialog.
'width% : the width of the progress bar in pixel. Should not be greater than the
'width of the dialog. If greater than the progress bar will be placed outside.
' Minimum width is 340, if provided less than it, then strips will made properly.
'value% : value can between 0 to 100. It represents how much portion of the
'progress bar will have green strip.
'active% :  It defines whether the progress bar is active or not. if given -1 then
'it is active
'###################################
'How shall I update the value of the progress bar ?
'Answer:
'Use aKUpdateProgress sub to update the value of the progress bar.
'Here's its syntax -
'aKUpdateProgress progressHandle%,newValue%
'progressHandle% : the handle of the progress bar.
'newValue% : the new value to be updated. Whether it is more than old value
'or less, the green stripp covering portion of progress bar will be
'updated.
DO
    'always use these two subs at the top of a loop with your dialog handle.
    aKCheck progressDialog
    aKUpdate progressDialog
    'checking if the user click run button
    IF aKClick(progressDialog, aKButton, runBtn) THEN
        'to prevent user next click we'll erase this object
        aKEraseObject progressDialog, aKButton, runBtn
        'and add progress bar
        start = -1
    END IF
    'updating our progress bar
    IF start AND p < 100 THEN
        p = p + 10
        akUpdateProgress progress, p
        'some delay. not neccessary
        _DELAY .2
    END IF
    _DISPLAY
    'checking if our dialog is closed by the user.
    IF aKDialogClose(progressDialog) THEN EXIT DO
LOOP

'$include:'akframework\akframework_method.bi'


demo/demo_colorpicker.bas
IF _DIREXISTS("akframework") THEN
    CHDIR "akframework\demo"
ELSEIF _DIREXISTS("demo") THEN
    CHDIR "demo"
END IF
'$include:'akframework\akframework_global.bi'
SCREEN _NEWIMAGE(800, 700, 32)

'setting our background
_PUTIMAGE , _LOADIMAGE("bg.jpg")
'setting window title
_TITLE "ColorPicker Demo"
'creating a dialog handle of colorpicker
colorpicker = aKNewdialog("Color Picker Dialog Demo", 400, 300)
'creating gradient based image from which the user
'will choose color.
gradient& = _NEWIMAGE(200, 200, 32)
_DEST gradient&
LINE (0, 0)-(200, 100), _RGB(255, 255, 255), BF
LINE (0, 101)-(200, 200), _RGB(0, 0, 0), BF
s = 101: a = 255
DO UNTIL s >= 200
    LINE (0, s)-(200, s), _RGBA(255, 255, 255, a)
    s = s + 1: a = a - 3
LOOP
'change it's value so that another gradient will be created
col& = _RGB(255, 0, 0)
s = 0: a = 0
DO UNTIL s > 100
    LINE (0, s)-(200, s), _RGBA(_RED(col&), _GREEN(col&), _BLUE(col&), a)
    s = s + 1: a = a + 3
LOOP
s = 101: a = 255
DO UNTIL s >= 200
    LINE (0, s)-(200, s), _RGBA(_RED(col&), _GREEN(col&), _BLUE(col&), a)
    s = s + 1: a = a - 3
LOOP
'creating a picture box with our gradient image handle
GrdImage = aKAddPicture(colorpicker, 10, 10, 200, 200, gradient&)
'adding a colorbar image which has raiblow color
bar = aKAddPicture(colorpicker, 10, 230, 175, 30, _LOADIMAGE("colorpicker.png"))
'adding a panel in which RGB values will be display.
rgbPanel = aKAddPanel(colorpicker, 220, 15, 170, 110, "RGB Values")
'now we will add three lables and numeric up-down in it which will
'show the rgb value of current color.
label1 = aKAddlabel(colorpicker, "Red", 225, 34)
label2 = aKAddlabel(colorpicker, "Green", 225, 64)
label3 = aKAddlabel(colorpicker, "Blue", 225, 94)
redValue = aKAddNumericUpDown(colorpicker, _RED(col&), 300, 30, 25)
greenValue = aKAddNumericUpDown(colorpicker, _GREEN(col&), 300, 60, 25)
blueValue = aKAddNumericUpDown(colorpicker, _BLUE(col&), 300, 90, 25)
'adding current color picture box
current& = _NEWIMAGE(150, 40, 32)
_DEST current&
PAINT (0, 0), col&
curPanel = aKAddPanel(colorpicker, 220, 137, 170, 60, "Current Color")
curImage = aKAddPicture(colorpicker, 230, 150, 150, 40, current&)
'and finally adding choose and cancel buttons
chooseBtn = aKAddButton(colorpicker, "Choose", 220, 235)
cancelBtn = aKAddButton(colorpicker, "Cancel", 300, 235)
'hurray!! we have created all the objects of our dialog
'they are as follows :-
' 3 Lables
' 3 Picture Boxes
' 3 Numeric Up-Down
' 2 Panels
' and 2 Buttons

'setting our drawing page to 0
_DEST 0
DO
    'always use these two subs at the top the loop with your dialog handle.
    aKCheck colorpicker
    aKUpdate colorpicker
    _DISPLAY
    'checking if the user click cancel button
    IF aKClick(colorpicker, aKButton, cancelBtn) THEN
        'we have to close the dialog
        aKHideDialog colorpicker
        END
    END IF
    'checking if the user click choose button
    IF aKClick(colorpicker, aKButton, chooseBtn) THEN
        'now, we'll show the color that the user has selected
        r = VAL(aKGetValue(aKNumericUpDown, redValue))
        g = VAL(aKGetValue(aKNumericUpDown, greenValue))
        b = VAL(aKGetValue(aKNumericUpDown, blueValue))
        aKHideDialog colorpicker
        COLOR _RGB(r, g, b)
        PRINT "You have choosen RGB("; r; ", "; g; " ,"; b; ")"
        END
    END IF
    'checking if the rgb values in the panel should not be more
    'than 255
    IF VAL(aKGetValue(aKNumericUpDown, redValue)) > 255 THEN
        'we'll set it's value to 255 by using aKSetValue method.
        aKSetValue aKNumericUpDown, redValue, "255"
        'also we have to update the content of the dialog
        aKDrawObject colorpicker, aKNumericUpDown, redValue
    END IF
    IF VAL(aKGetValue(aKNumericUpDown, greenValue)) > 255 THEN
        aKSetValue aKNumericUpDown, greenValue, "255"
        aKDrawObject colorpicker, aKNumericUpDown, greenValue
    END IF
    IF VAL(aKGetValue(aKNumericUpDown, blueValue)) > 255 THEN
        aKSetValue aKNumericUpDown, blueValue, "255"
        aKDrawObject colorpicker, aKNumericUpDown, blueValue
    END IF
    'and also not less than 0 by the user
    IF VAL(aKGetValue(aKNumericUpDown, redValue)) < 0 THEN
        aKSetValue aKNumericUpDown, redValue, "0"
        aKDrawObject colorpicker, aKNumericUpDown, redValue
    END IF
    IF VAL(aKGetValue(aKNumericUpDown, greenValue)) < 0 THEN
        aKSetValue aKNumericUpDown, greenValue, "0"
        aKDrawObject colorpicker, aKNumericUpDown, greenValue
    END IF
    IF VAL(aKGetValue(aKNumericUpDown, blueValue)) < 0 THEN
        aKSetValue aKNumericUpDown, blueValue, "0"
        aKDrawObject colorpicker, aKNumericUpDown, blueValue
    END IF
    'checking if the any Numeric Up-down of the dialog are
    'clicked means there value has been changed
    IF aKAnyClick(colorpicker, aKNumericUpDown) THEN
        'now we have to update our current color picture box
        r = VAL(aKGetValue(aKNumericUpDown, redValue))
        g = VAL(aKGetValue(aKNumericUpDown, greenValue))
        b = VAL(aKGetValue(aKNumericUpDown, blueValue))
        current& = _NEWIMAGE(150, 40, 32)
        _DEST current&
        PAINT (0, 0), _RGB(r, g, b)
        _DEST 0
        aKSetPicture curImage, current&
        _FREEIMAGE current&
        aKDrawObject colorpicker, aKPicture, curImage
    END IF
    'checking if the gradient picture box is clicked by the user
    'then update current color as well as rgb values
    IF aKClick(colorpicker, aKPicture, GrdImage) THEN
        'note that Mouse.x, Mouse.y, Mouse.Lclick, Mouse.Rclick are the variables
        'defined in the akframework_global.bi
        col& = POINT(Mouse.x, Mouse.y)
        aKSetValue aKNumericUpDown, redValue, STR$(_RED(col&))
        aKSetValue aKNumericUpDown, greenValue, STR$(_GREEN(col&))
        aKSetValue aKNumericUpDown, blueValue, STR$(_BLUE(col&))
        current& = _NEWIMAGE(150, 40, 32)
        _DEST current&
        PAINT (0, 0), _RGB(_RED(col&), _GREEN(col&), _BLUE(col&))
        _DEST 0
        aKSetPicture curImage, current&
        _FREEIMAGE current&
        aKDrawObject colorpicker, aKPicture, curImage
        aKDrawObject colorpicker, aKNumericUpDown, redValue
        aKDrawObject colorpicker, aKNumericUpDown, greenValue
        aKDrawObject colorpicker, aKNumericUpDown, blueValue
    END IF
    'and last checking that the bar has been clicked by the user
    'so we have to update our color gradient
    IF aKClick(colorpicker, aKPicture, bar) THEN
        'recreating our gradient
        col& = POINT(Mouse.x, Mouse.y)
        gradient& = _NEWIMAGE(200, 200, 32)
        _DEST gradient&
        LINE (0, 0)-(200, 100), _RGB(255, 255, 255), BF
        LINE (0, 101)-(200, 200), _RGB(0, 0, 0), BF
        s = 101: a = 255
        DO UNTIL s >= 200
            LINE (0, s)-(200, s), _RGBA(255, 255, 255, a)
            s = s + 1: a = a - 3
        LOOP
        s = 0: a = 0
        DO UNTIL s > 100
            LINE (0, s)-(200, s), _RGBA(_RED(col&), _GREEN(col&), _BLUE(col&), a)
            s = s + 1: a = a + 3
        LOOP
        s = 101: a = 255
        DO UNTIL s >= 200
            LINE (0, s)-(200, s), _RGBA(_RED(col&), _GREEN(col&), _BLUE(col&), a)
            s = s + 1: a = a - 3
        LOOP
        _DEST 0
        aKSetPicture GrdImage, gradient&
        _FREEIMAGE gradient&
        aKDrawObject colorpicker, aKPicture, GrdImage
    END IF
    'checking if the dialog is closed by the user
    IF aKDialogClose(colorpicker) THEN EXIT DO
LOOP

'$include:'akframework\akframework_method.bi'
